# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/base/nsFocusManager.cpp
# Commit: d4411b38cae7
# Full Hash: d4411b38cae7216ffc57a9b8b246806acf0a3717
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2023-12-24 09:09:23
# Regressor Bug: 779684
# File Overlap Count: 1
# Description:
#   Bug 779684 - part 1: Get rid of `nsIFrameEnumerator` interface r=emilio
#   
#   It's inherited only by `nsFrameIterator`, and `nsFrameIterator` can be declared
#   in a header file. So, the interface is not required and removing it can avoid
#   virtual calls.
# ==============================================================================

diff -r bda7c9e66417 -r d4411b38cae7 dom/base/nsFocusManager.cpp
--- a/dom/base/nsFocusManager.cpp	Sat Dec 23 19:02:04 2023 +0000
+++ b/dom/base/nsFocusManager.cpp	Sun Dec 24 06:01:19 2023 +0000
@@ -3221,12 +3221,12 @@
       domSelection->IsCollapsed()) {
     nsIFrame* startFrame = start->GetPrimaryFrame();
     // Yes, indeed we were at the end of the last node
-    nsCOMPtr<nsIFrameEnumerator> frameTraversal;
+    RefPtr<nsFrameIterator> frameIterator;
     nsIFrame* limiter =
         domSelection && domSelection->GetAncestorLimiter()
             ? domSelection->GetAncestorLimiter()->GetPrimaryFrame()
             : nullptr;
-    MOZ_TRY(NS_NewFrameTraversal(getter_AddRefs(frameTraversal), presContext,
+    MOZ_TRY(NS_NewFrameTraversal(getter_AddRefs(frameIterator), presContext,
                                  startFrame, eLeaf,
                                  false,  // aVisual
                                  false,  // aLockInScrollView
@@ -3241,8 +3241,8 @@
       // Continue getting the next frame until the primary content for the
       // frame we are on changes - we don't want to be stuck in the same
       // place
-      frameTraversal->Next();
-      newCaretFrame = static_cast<nsIFrame*>(frameTraversal->CurrentItem());
+      frameIterator->Next();
+      newCaretFrame = frameIterator->CurrentItem();
       if (!newCaretFrame) {
         break;
       }
@@ -4193,13 +4193,13 @@
       getNextFrame = false;
     }
 
-    nsCOMPtr<nsIFrameEnumerator> frameTraversal;
+    RefPtr<nsFrameIterator> frameIterator;
     if (frame) {
       // For tab navigation, pass false for aSkipPopupChecks so that we don't
       // iterate into or out of a popup. For document naviation pass true to
       // ignore these boundaries.
       nsresult rv = NS_NewFrameTraversal(
-          getter_AddRefs(frameTraversal), presContext, frame, ePreOrder,
+          getter_AddRefs(frameIterator), presContext, frame, ePreOrder,
           false,                  // aVisual
           false,                  // aLockInScrollView
           true,                   // aFollowOOFs
@@ -4209,18 +4209,18 @@
 
       if (iterStartContent == aRootContent) {
         if (!aForward) {
-          frameTraversal->Last();
+          frameIterator->Last();
         } else if (aRootContent->IsFocusableWithoutStyle()) {
-          frameTraversal->Next();
+          frameIterator->Next();
         }
-        frame = frameTraversal->CurrentItem();
+        frame = frameIterator->CurrentItem();
       } else if (getNextFrame &&
                  (!iterStartContent ||
                   !iterStartContent->IsHTMLElement(nsGkAtoms::area))) {
         // Need to do special check in case we're in an imagemap which has
         // multiple content nodes per frame, so don't skip over the starting
         // frame.
-        frame = frameTraversal->Traverse(aForward);
+        frame = frameIterator->Traverse(aForward);
       }
     }
 
@@ -4242,11 +4242,11 @@
         // We're within non-document scope, continue.
         do {
           if (aForward) {
-            frameTraversal->Next();
+            frameIterator->Next();
           } else {
-            frameTraversal->Prev();
+            frameIterator->Prev();
           }
-          frame = static_cast<nsIFrame*>(frameTraversal->CurrentItem());
+          frame = frameIterator->CurrentItem();
           // For the usage of GetPrevContinuation, see the comment
           // at the end of while (frame) loop.
         } while (frame && frame->GetPrevContinuation());
@@ -4506,11 +4506,11 @@
       // again.
       do {
         if (aForward) {
-          frameTraversal->Next();
+          frameIterator->Next();
         } else {
-          frameTraversal->Prev();
+          frameIterator->Prev();
         }
-        frame = static_cast<nsIFrame*>(frameTraversal->CurrentItem());
+        frame = frameIterator->CurrentItem();
       } while (frame && frame->GetPrevContinuation());
     }
 